src/segment-loader.js

```diff
@@ -399,25 +403,39 @@ export default class SegmentLoader extends videojs.EventTarget {
 
       // if the video has not yet played only, and we already have
       // one segment downloaded do nothing
-      if (!this.hasPlayed_() && bufferedTime >= 1) {
+      if (!hasPlayed && bufferedTime >= 1) {
         return null;
       }
 
       // if there is plenty of content buffered, and the video has
       // been played before relax for awhile
-      if (this.hasPlayed_() && bufferedTime >= Config.GOAL_BUFFER_LENGTH) {
+      if (hasPlayed && bufferedTime >= Config.GOAL_BUFFER_LENGTH) {
         return null;
       }
       mediaIndex = getMediaIndexForTime(playlist,
-                                        currentBufferedEnd + this.timeCorrection_,
-                                        this.expired_);
+                                        currentBufferedEnd + timeCorrection,
+                                        expired);
     }
 
-    if (mediaIndex < 0 || mediaIndex === playlist.segments.length) {
+    if (mediaIndex === playlist.segments.length) {
+      if (playlist.endList &&
+          buffered.length) {
+        let endOfBuffer = buffered.end(buffered.length - 1);
+        let lastSegment = playlist.segments[playlist.segments.length - 1];
+
+        if (lastSegment.end &&
+            lastSegment.end < endOfBuffer + Ranges.TIME_FUDGE_FACTOR) {
+          return { isEndOfStream: true };
+        }
+      }
+      return null;
+    }
```

Comment:
Can we add the slight differences from this code to the code above for `detectEndOfStream` and move this check from here to `fillBuffer_`?